internal/app/orchestrator/orchestrator.go

```diff
@@ -95,16 +115,197 @@ func New(ctx context.Context,
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *discovery.DiscoveryResponse,
+	done <-chan bool,
+	shutdownUpstream func(),
+) {
+	for {
+		select {
+		case x, more := <-responseChannel:
+			if !more {
+				// A problem occurred fetching the response upstream, log retry.
+				// TODO implement retry/back-off logic on error scenario.
+				// https://github.com/envoyproxy/xds-relay/issues/68
+				o.logger.With("key", aggregatedKey).Error(ctx, "upstream error")
+				return
+			}
+			// Cache the response.
+			_, err := o.cache.SetResponse(aggregatedKey, *x)
+			if err != nil {
+				// TODO if set fails, we may need to retry upstream as well.
+				// Currently the fallback is to rely on a future response, but
+				// that probably isn't ideal.
+				// https://github.com/envoyproxy/xds-relay/issues/70
+				//
+				// If we fail to cache the new response, log and return the old one.
+				o.logger.With("err", err).With("key", aggregatedKey).
+					Error(ctx, "Failed to cache the response")
+			}
+
+			// Get downstream watches and fan out.
+			// We retrieve from cache rather than directly fanning out the
+			// newly received response because the cache does additional
+			// resource serialization.
+			cached, err := o.cache.Fetch(aggregatedKey)
+			if err != nil {
+				o.logger.With("err", err).With("key", aggregatedKey).Error(ctx, "cache fetch failed")
+				// Can't do anything because we don't know who the watches
+				// are. Drop the response.
+			} else {
+				if cached == nil || cached.Resp == nil {
+					// If cache is empty, there is nothing to fan out.
+					// Error. Sanity check. Shouldn't ever reach this since we
+					// just set the response, but it's a rare scenario that can
+					// happen if the cache TTL is set very short.
+					o.logger.With("key", aggregatedKey).Error(ctx, "attempted to fan out with no cached response")
+				} else {
+					// Goldenpath.
+					o.fanout(cached.Resp, cached.Requests, aggregatedKey)
+				}
+			}
+		case <-done:
+			// Exit when signaled that the stream has closed.
+			shutdownUpstream()
+			return
+		}
+	}
+}
+
+// fanout pushes the response to the response channels of all open downstream
+// watches in parallel.
+func (o *orchestrator) fanout(resp *cache.Response, watchers map[*gcp.Request]bool, aggregatedKey string) {
+	var wg sync.WaitGroup
+	for watch := range watchers {
+		wg.Add(1)
+		go func(watch *gcp.Request) {
+			defer wg.Done()
+			if channel, ok := o.downstreamResponseMap.get(watch); ok {
+				select {
+				case channel <- convertToGcpResponse(resp, *watch):
+					break
+				default:
```

Comment:
How comment is this expected to happen? It looks like this channel is buffered as 1. if there is spam of updates, maybe it is fine to drop?

Reply:
There's more discussion for this [here](https://github.com/envoyproxy/xds-relay/pull/53/files#r417671109).

The idea behind this was that a downstream receiver could be blocked for any reason. Initially I had went with a timeout, but in discussions with @jyotimahapatra we realized that it would imply the slowest sidecar would block new updates from being sent to other sidecars, due to the use of WaitGroups. We can't get rid of WaitGroups because more recent updates might get overriden by a slower update if updates were bursty.

The buffer is necessary in order to give time for the receiver to initialize. I realized in [this test](https://github.com/envoyproxy/xds-relay/pull/53/files#diff-2c2ed0f611d2764f6a076c08ee2b18f7R254) that the second watch (for the same aggregated key) would fall to the `default` case if the channel was unbuffered.

If there is a spam of updates, I think it's reasonable to drop since only one of those updates will be applied downstream anyway, but we do lose the the guarantee that it is the "latest" update. Could move back to the configured timeout model if this is a concern, but I put some of those downsides ^.

Reply:
My thoughts are to move forward with the default for now. I'll document this. If it is a problem in practice, then we can revisit the timeout model or other approaches.

Reply:
sounds good to me.

Reply:
Yeah in the future, you can do what Envoymanager does and drop the oldest and enqueue the newest. That has been working quite well for us.